### PR TITLE
Fix VMRay missing process data

### LINKED ISSUE #2394: KeyError:3 while running capa on VMRay analysis archive
### Description

While running capa 7.3.0 on a VMRay analysis archive, capa crashed and provided a stack trace for troubleshooting.

### Steps to Reproduce

- Obtain VMRay sample archive for sample `2f8a79b12a7a989ac7e5f6ec65050036588a92e65aeb6841e08dc228ff0e21b4` (attached)
- Run `./capa -d -f vmray 2f8a79b12a7a989ac7e5f6ec65050036588a92e65aeb6841e08dc228ff0e21b4_analysis_archive.zip`
- Observe output

**Expected behavior:**

I expected a capability report.

**Actual behavior:**

The capa process crashed with stack trace:

```
DEBUG:capa:--------------------------------------------------------------------------------
DEBUG:capa: Using default embedded rules.
DEBUG:capa: To provide your own rules, use the form:
DEBUG:capa:
DEBUG:capa:     `capa.exe -r ./path/to/rules/  /path/to/mal.exe`.
DEBUG:capa:
DEBUG:capa: You can see the current default rule set here:
DEBUG:capa:
DEBUG:capa:     https://github.com/mandiant/capa-rules
DEBUG:capa:--------------------------------------------------------------------------------
DEBUG:capa.rules:reading rules from directory /tmp/_MEIoxSNmT/rules
DEBUG:capa.rules.cache:loading rule set from cache: /tmp/_MEIoxSNmT/cache/capa-df400217.cache
DEBUG:capa:successfully loaded 901 rules
DEBUG:capa.features.extractors.vmray:file_type: Windows Exe (x86-32), file_path: internal/static_analyses/2f8a79b12a7a989ac7e5f6ec65050036588a92e65aeb6841e08dc228ff0e21b4/objects/files/2f8a7
9b12a7a989ac7e5f6ec65050036588a92e65aeb6841e08dc228ff0e21b4
Traceback (most recent call last):
  File "main.py", line 1094, in <module>
  File "main.py", line 956, in main
  File "main.py", line 708, in get_file_extractors_from_cli
  File "loader.py", line 390, in get_file_extractors
  File "features/extractors/vmray/extractor.py", line 122, in from_zipfile
  File "features/extractors/vmray/__init__.py", line 87, in __init__
  File "features/extractors/vmray/__init__.py", line 141, in _compute_process_threads
  File "features/extractors/vmray/__init__.py", line 161, in get_process_os_pid
KeyError: 3
[PYI-1584:ERROR] Failed to execute script 'main' due to unhandled exception!
```

### Versions

capa 7.3.0
5.15.0-1047-aws 20.04.1-Ubuntu x86_64

### Additional Information

See attached analysis archive for troubleshooting
[2f8a79b12a7a989ac7e5f6ec65050036588a92e65aeb6841e08dc228ff0e21b4_analysis_archive.zip](https://github.com/user-attachments/files/17105304/2f8a79b12a7a989ac7e5f6ec65050036588a92e65aeb6841e08dc228ff0e21b4_analysis_archive.zip)

### PR DESCRIPTION
**Pull request by mr-tz**
closes #2394

### Checklist

- [ ] No CHANGELOG update needed

- [ ] No new tests needed

- [ ] No documentation update needed

### CODE REVIEW

**Comment by williballenthin** [comment]
looks reasonable, i'll trust the tests.

**Review by williballenthin — APPROVED** [review]

**Comment by mike-hunhoff** [comment]
@mr-tz thanks for the initial fix. This made me realize that we should be using VMRay's thread and process monitor IDs instead of tid, pid, and ppid combinations as we can't guarantee these combinations are unique, e.g. reused ppid and pid combination, while it appears that the thread and process monitor IDs are unique, and thus combinations of these should be unique, e.g. for extracting function calls.

File: capa/features/extractors/vmray/__init__.py [hunk 1]
```diff
@@ -128,41 +147,42 @@ def _compute_sections(self):
...
-            else:
-                self.os_pid_by_monitor_id[monitor_process.process_id] = monitor_process.os_pid
-
-    def _compute_process_threads(self):
-        # logs/flog.xml appears to be the only file that contains thread-related data
-        # so we use it here to map processes to threads
-        for function_call in self.flog.analysis.function_calls:
-            pid: int = self.get_process_os_pid(function_call.process_id)  # flog.xml uses process monitor ID, not OS PID
-            tid: int = function_call.thread_id
+            if monitor_process.process_id not in self.monitor_processes.keys():
```
Comment (mr-tz) [inline capa/features/extractors/vmray/__init__.py:164]: should we perform some sanity checks on the existing data, i.e., that the IDs line up?

File: capa/features/extractors/vmray/__init__.py [hunk 2]
```diff
@@ -128,41 +147,42 @@ def _compute_sections(self):
...
                 self.sections[elffile_section.header.sh_addr] = elffile_section.header.sh_name
 
-    def _compute_process_ids(self):
+    def _compute_monitor_processes(self):
         for process in self.sv2.processes.values():
-            # we expect VMRay's monitor IDs to be unique, but OS PIDs may be reused
-            assert process.monitor_id not in self.os_pid_by_monitor_id.keys()
-            self.os_pid_by_monitor_id[process.monitor_id] = process.os_pid
+            # we expect monitor IDs to be unique
+            assert process.monitor_id not in self.monitor_processes.keys()
```
Comment (mr-tz) [inline capa/features/extractors/vmray/__init__.py:153]: don't need `.keys()` here and in similar lines below

File: capa/features/extractors/vmray/__init__.py [hunk 3]
```diff
@@ -128,41 +147,42 @@ def _compute_sections(self):
...
+            # we expect monitor IDs to be unique
+            assert monitor_thread.thread_id not in self.monitor_threads.keys()
+
+            self.monitor_threads[monitor_thread.thread_id] = VMRayMonitorThread(
+                monitor_thread.os_tid, monitor_thread.thread_id, monitor_thread.process_id
+            )
 
-            assert isinstance(pid, int)
-            assert isinstance(tid, int)
+            # we expect 1 monitor thread per monitor process
```
Comment (mr-tz) [inline capa/features/extractors/vmray/__init__.py:181]: this comment confuses me (only 1 thread per process?) and the following 2 lines appear unrelated?

File: capa/features/extractors/vmray/__init__.py [hunk 4]
```diff
@@ -45,9 +61,11 @@ def __init__(self, zipfile_path: Path):
         self.exports: Dict[int, str] = {}
         self.imports: Dict[int, Tuple[str, str]] = {}
         self.sections: Dict[int, str] = {}
-        self.os_pid_by_monitor_id: Dict[int, int] = {}
-        self.tids_by_pid: Dict[int, List[int]] = defaultdict(list)
-        self.process_calls: Dict[int, Dict[int, List[FunctionCall]]] = defaultdict(lambda: defaultdict(list))
+        self.monitor_processes: Dict[int, VMRayMonitorProcess] = {}
+        self.monitor_threads: Dict[int, VMRayMonitorThread] = {}
+        self.monitor_threads_by_monitor_process: Dict[int, List[int]] = defaultdict(list)
+        self.monitor_process_calls: Dict[int, Dict[int, List[FunctionCall]]] = defaultdict(lambda: defaultdict(list))
```
Comment (mr-tz) [inline capa/features/extractors/vmray/__init__.py:67]: I think two respective comments here would be useful to know what maps to what (pid and tid)

File: capa/features/extractors/vmray/__init__.py [hunk 5]
```diff
@@ -128,41 +147,42 @@ def _compute_sections(self):
             for elffile_section in self.sample_file_static_data.elf.sections:
                 self.sections[elffile_section.header.sh_addr] = elffile_section.header.sh_name
 
-    def _compute_process_ids(self):
+    def _compute_monitor_processes(self):
         for process in self.sv2.processes.values():
```
Comment (mr-tz) [inline capa/features/extractors/vmray/__init__.py:151]: do we even need to process the summary json here now?

File: CHANGELOG.md [hunk 6]
```diff
@@ -21,6 +21,7 @@
 
 - use Python 3.12 to build extra standalone build on Linux #2383 @williballenthin
 - bump minimum Python version to 3.8.1 to satisfy uv #2387 @williballenthin
+- collect more process information from flog.xml #2394 @mr-tz
```
Comment (williballenthin) [inline CHANGELOG.md:24]:
```suggestion
- vmray: collect more process information from flog.xml #2394 @mr-tz
```

File: capa/features/extractors/vmray/__init__.py [hunk 7]
```diff
@@ -21,6 +22,21 @@
 SUPPORTED_FLOG_VERSIONS = ("2",)
 
 
+@dataclass
+class VMRayMonitorThread:
+    tid: int  # thread ID assigned by OS
+    monitor_id: int  # unique ID assigned to thread by VMRay
+    process_monitor_id: int  # unqiue ID assigned to containing process by VMRay
```
Comment (williballenthin) [inline capa/features/extractors/vmray/__init__.py:29]: thank you for the comments, very helpful

**Comment by mr-tz** [comment]
I ran this on a few random private analysis archives and didn't encounter any failures. We should find more public samples to add to the test corpus.

**Review by williballenthin — APPROVED** [review]
should we address #2361 (add fields to unique pid/ppid) here, while we're at it? or leave that for a PR immediately following this?

File: capa/features/extractors/vmray/__init__.py (hunk 1 above)
Comment (mike-hunhoff) [reply capa/features/extractors/vmray/__init__.py:171]: https://github.com/mandiant/capa/pull/2396/commits/df878c78576ae66444bb0b49257d1f27ba4d5d91

File: capa/features/extractors/vmray/__init__.py (hunk 2 above)
Comment (mike-hunhoff) [reply capa/features/extractors/vmray/__init__.py]: https://github.com/mandiant/capa/pull/2396/commits/3cc48f18744baed375e09bd82bfc67478e531280

Comment (mike-hunhoff) [reply capa/features/extractors/vmray/__init__.py]: Good catch!

File: capa/features/extractors/vmray/__init__.py (hunk 3 above)
Comment (mike-hunhoff) [reply capa/features/extractors/vmray/__init__.py:186]: Ah yes I can see how this is comment is confusing. I intended to communicate that we expect each monitor thread ID that is associated with each monitor process ID to be unique, e.g. the monitor thread ID 10 should not be associated with monitor process ID 1 more than once. This is because it appears that monitor process/thread IDs are incrementally unique for each process/thread monitored by VMRay. I'll attempt to clarify this better with an updated comment.

File: capa/features/extractors/vmray/__init__.py (hunk 4 above)
Comment (mike-hunhoff) [reply capa/features/extractors/vmray/__init__.py]: https://github.com/mandiant/capa/pull/2396/commits/d331f22639fc83977d8707408498431336f50872

File: capa/features/extractors/vmray/__init__.py (hunk 5 above)
Comment (mike-hunhoff) [reply capa/features/extractors/vmray/__init__.py:155]: We currently don't emit features at the process level, only function calls which are recorded in `flog.xml`. So, as of now, the answer is no we probably don't, but to future proof the code we'll need to process both `SummaryV2.json` and `flog.xml` to achieve full coverage at the process scope, so I suggest we keep it. Please reopen if you feel strongly otherwise.

**Comment by mike-hunhoff** [comment]
> should we address #2361 (add fields to unique pid/ppid) here, while we're at it? or leave that for a PR immediately following this?

We should address #2361 in a separate PR

**Comment by mr-tz** [comment]
Thanks for the additions, @mike-hunhoff!
